Dynamic screenshots

murugapl opened the pull request.

murugapl committed:
Initial playwright implementation

- Login to local setup
- Navigate to target URL
- Screenshot page

murugapl committed:
Parameterised image saving

- JSON file for image metadata

murugapl committed:
Screenshots for custom elements

- Find element handle and size details
- Custom functions for processing images
- Refactor

murugapl committed:
Screenshot consent response

- Find and process radio buttons

murugapl committed:
Screenshot full pages

- util func to remove footer from full page screenshot

murugapl committed:
Process rest of images

- refactor

murugapl committed:
Bug fixing

murugapl committed:
Minor improvements

murugapl committed:
README fixes

murugapl committed:
Move requirements.txt

murugapl committed:
Make README clearer

murugapl committed:
Bug fixes
- environments README
- extra image in generate_images.py

murugapl committed:
Remove unused variable

alistair-white-horne-tng committed:
Change the output directory to app/assets/images

Also remove descriptions from the images
And add TODOs to the README.md

murugapl committed:
Update README.md

murugapl committed:
Fix error in readme

mikejamesthompson commented on dynamic_screenshots/process_images.py:
```diff
@@ -0,0 +1,101 @@
+    child_row = child_key.locator("..")
+    child_value = await child_row.locator("dd").inner_text()
+    await page.get_by_label("Name", exact=True).fill(child_value)
+    time.sleep(5)
```
With something like Playwright, I wouldn't expect you to have to explicitly sleep to wait for something to happen.

Can you use something like `page.expect_response`? https://playwright.dev/python/docs/api/class-page#page-wait-for-event

murugapl replied:
There isn't anything direct to expect I don't think, so I have swapped it to page.wait_for_timeout from playwright.

mikejamesthompson commented on dynamic_screenshots/generate_images.py:
```diff
@@ -0,0 +1,67 @@
+            page = await context.new_page()
+
+            # Login to the application
+            if "login" in image.keys():
```
If this code path isn't being used right now, then you can safely remove it in another commit -- it'll be in version control. Best not to keep code paths that aren't being used as it can be confusing.

mikejamesthompson commented on dynamic_screenshots/images.py:
```diff
@@ -0,0 +1,201 @@
+    },
+    {
+        "image_name": "consent-response-paper.png",
+        "url_extension": "sessions/1/consents/given",
```
As with my comment on the readme, I suggest you rename `url_extension` to `path`.

mikejamesthompson reviewed (changes requested):
This is looking good, but I'm having issues running it as it is because the session URLs have changed and even deleting the database the IDs used for consent and school moves don't match up to what I see locally.

Apart from that, I've added some other suggestions, including some quite picky ones 😃

murugapl commented on dynamic_screenshots/generate_images.py:
```diff
@@ -0,0 +1,67 @@
+    await page.get_by_label("Password", exact=True).fill(password)
+    await page.get_by_role("button", name="Log in").click()
+    await page.wait_for_url("**/users/organisations/new", timeout=1000)
+    await page.pause()
```
This was added originally as there was an issue when the login took longer than expected so the next line in the script could not find the text it was looking for. I have removed it now as I agree it is too specific a solution in this case, and we can assume that if the script is run again, it will work.

murugapl commented on dynamic_screenshots/images.py:
```diff
@@ -0,0 +1,201 @@
+    },
+    {
+        "image_name": "session-attendance.png",
+        "url_extension": "sessions/1/attendances/unregistered?sort=name&direction=asc",
```
That is strange it still works for me with `sessions/1/...`. Which version are you running?

murugapl replied:
Ah seems to have been a change made a while ago, not sure why we had a different version. Fixed now - all urls are `sessions/session-1 `etc

murugapl committed:
Refactorings

- rename 'url_extension' to 'path'
- Improve readability of README
- Make login function more uniform

murugapl commented on dynamic_screenshots/generate_images.py:
```diff
@@ -0,0 +1,67 @@
+
+        full_page = image["full_page"] if "full_page" in image.keys() else False
+
+        async with async_playwright() as p:
```
I have implemented this now and it is indeed quicker. However, it means we lose the ability to login as a superuser, as we only login once at the start now- this was the reason we did not do it this way round before. I think it is fine to do this given that we do not need superuser screenshots at the moment.

murugapl committed:
Refactor screenshot taking order
- Launch playwright and login once
- quicker overall

murugapl committed:
Swap time.sleep for playwright.wait_for_timeout, remove commented code

murugapl committed:
Rename image paths to fit new convention

murugapl committed:
Merge pull request #2 from NHSDigital/main

Merge in 2.0 updates to main

murugapl committed:
Merge pull request #4 from NHSDigital/main

Make main branch up to date

murugapl committed:
Update screenshots for v2.1

Commit:
Merge branch 'main' of https://github.com/murugapl/manage-vaccinations-in-schools-guidance into dynamic-screenshots

murugapl committed:
Add v2.1 extra screenshots

murugapl committed:
Update README.md
